spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchScan.java

```diff
@@ -208,6 +209,38 @@ public Statistics estimateStatistics() {
     return new Stats(sizeInBytes, numRows);
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    SparkBatchScan that = (SparkBatchScan) o;
+    // compare Spark schemas to ignore field ids
+    StructType thisSchema = SparkSchemaUtil.convert(expectedSchema);
+    StructType thatSchema = SparkSchemaUtil.convert(that.expectedSchema);
```

Comment:
Oops, I'm also missing an update that uses `readSchema()` and a lazy variable to avoid converting multiple times.